static/app/components/arithmeticBuilder/validator.tsx

```diff
@@ -19,106 +21,172 @@ export function validateTokens(tokens: Token[]): boolean {
   return validator.isValid();
 }
 
-enum Part {
-  OPEN_PAREN = '(',
-  OPERATOR = 'o',
-  EXPRESSION = 'e',
+export function computeNextAllowedTokenKinds(tokens: Array<Token | null>): TokenKind[][] {
+  const validator = new ExpressionValidator();
+  const tokenKinds = [];
+
+  for (const token of tokens) {
+    if (!defined(token)) {
+      tokenKinds.push([]);
+    } else if (isTokenFreeText(token)) {
+      tokenKinds.push(validator.nextAllowedTokenKinds());
+    } else {
+      if (validator.push(token)) {
```

Comment:
comment suggestion: This actually pushes the token on to the validation stack if it's a valid token